network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+
+    // Collect the set of updated addresses.
+    let updates: Vec<ConnectivityRequest> = full_node_infos
+        .into_iter()
+        .map(|full_node| {
+            ConnectivityRequest::UpdateAddresses(
+                DiscoverySource::OnChain,
+                // TODO:  this seems wrong for the account address.  Is it shared with the validator?
+                full_node.account_address().clone(),
+                vec![NetworkAddress::try_from(
+                    &full_node.config().full_node_network_address.clone(),
+                )
+                .expect("WTF")],
+            )
+        })
+        .collect();
+
+    // Collect the EligibleNodes
+    updates.push(ConnectivityRequest::UpdateEligibleNodes(
+        full_node_infos
+            .into_iter()
+            .map(|full_node| {
+                (
+                    // TODO:  This seems like it is wrong and should be something else.  Or is the full_node account address the same as the validator?
+                    full_node.account_address().clone(),
+                    NetworkPublicKeys {
+                        identity_public_key: full_node
+                            .config()
+                            .full_node_network_identity_public_key
+                            .clone(),
+                        // TODO:  there are not separate keys for the full_node.  Are they therefore the same?
+                        // Note that signing_key type is Ed25519PublicKey
+                        // public key type is x25519::PublicKey
+                        signing_public_key: full_node
+                            .config()
+                            .full_node_network_identity_public_key
+                            .clone(),
+                    },
+                )
+            })
+            .collect(),
+    ));
+
+    updates
+}
+
+impl ConfigurationChangeListener {
+    pub fn new(conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>, role: RoleType) -> Self {
+        Self {
+            conn_mgr_reqs_tx,
+            role,
+        }
+    }
+
+    /// Processes a received OnChainConfigPayload.  Depending on role (Validator or FullNode), parses
+    /// the appropriate configuration changes and passes it to the ConnectionManager channel.
+    async fn process_payload(&mut self, payload: OnChainConfigPayload) {
+        let node_set: ValidatorSet = payload
+            .get()
+            .expect("failed to get ValidatorSet from payload");
+
+        let updates = match self.role {
+            RoleType::Validator => extract_validator_updates(node_set),
+            RoleType::FullNode => extract_full_node_updates(node_set),
+        };
+
+        info!(
+            "Update {} Network about new Node IDs",
+            self.role.to_string()
+        );
+        for mut update in updates {
+            let err_msg;
+            match &update {
+                ConnectivityRequest::UpdateAddresses(a, b, c) => {
```

Comment:
maybe clearer names here? should we consider putting this logging into the place where we formatted the objects? is this duplicative of what is happening within the connection manager?

Reply:
It is duplicate-esque to what is happening in ConnectivityManager, but there is an asynchronous data transition between the two logging points.  Failure in one should correlate with failure in the other, but does not necessarily if some other part of the system happens to be borked as well.